File: cooking-cost-system/backend/src/routes/reports.ts
import { Router, Request, Response } from 'express';
import { asyncHandler } from '../middleware/errorHandler';
import { Ingredient } from '../models/Ingredient';
import { Dish } from '../models/Dish';
import { CompletedFood } from '../models/CompletedFood';
import { getDatabase } from '../database';
import { logger } from '../utils/logger';

const router = Router();

// GET /api/reports/dashboard - ダッシュボード統計
router.get('/dashboard', asyncHandler(async (req: Request, res: Response) => {
    const db = getDatabase();

    // 基本統計
    const [totalIngredients] = await db.query('SELECT COUNT(*) as count FROM ingredients');
    const [totalDishes] = await db.query('SELECT COUNT(*) as count FROM dishes');
    const [totalFoods] = await db.query('SELECT COUNT(*) as count FROM completed_foods');

    // 平均利益率計算
    const [avgProfitRate] = await db.query(`
        SELECT AVG(CASE WHEN price > 0 THEN ((price - total_cost) / price) * 100 ELSE 0 END) as avg_profit_rate
        FROM completed_foods
        WHERE price IS NOT NULL AND price > 0
    `);

    // 総売上・原価・利益
    const [totals] = await db.query(`
        SELECT 
            COALESCE(SUM(price), 0) as total_revenue,
            COALESCE(SUM(total_cost), 0) as total_cost,
            COALESCE(SUM(price - total_cost), 0) as total_profit
        FROM completed_foods
        WHERE price IS NOT NULL
    `);

    // 最近の活動
    const recentActivity = await db.query(`
        (SELECT 'ingredient' as type, CONCAT('食材「', name, '」を追加') as message, created_at as timestamp
         FROM ingredients ORDER BY created_at DESC LIMIT 5)
        UNION ALL
        (SELECT 'dish' as type, CONCAT('料理「', name, '」を作成') as message, created_at as timestamp
         FROM dishes ORDER BY created_at DESC LIMIT 5)
        UNION ALL
        (SELECT 'food' as type, CONCAT('完成品「', name, '」を登録') as message, created_at as timestamp
         FROM completed_foods ORDER BY created_at DESC LIMIT 5)
        ORDER BY timestamp DESC LIMIT 10
    `);

    const dashboardData = {
        summary: {
            totalIngredients: totalIngredients[0]?.count || 0,
            totalDishes: totalDishes[0]?.count || 0,
            totalCompletedFoods: totalFoods[0]?.count || 0,
            avgProfitRate: parseFloat(avgProfitRate[0]?.avg_profit_rate || '0').toFixed(2),
            totalRevenue: parseFloat(totals[0]?.total_revenue || '0'),
            totalCost: parseFloat(totals[0]?.total_cost || '0'),
            totalProfit: parseFloat(totals[0]?.total_profit || '0'),
        },
        recentActivity: recentActivity.map((activity: any, index: number) => ({
            id: index + 1,
            type: activity.type,
            message: activity.message,
            timestamp: activity.timestamp,
        })),
    };

    logger.info('Generated dashboard report');

    res.json({
        success: true,
        data: dashboardData,
        timestamp: new Date().toISOString(),
    });
}));

// GET /api/reports/genre-stats - ジャンル統計
router.get('/genre-stats', asyncHandler(async (req: Request, res: Response) => {
    const ingredientStats = await Ingredient.getGenreStatistics();
    const dishStats = await Dish.getGenreStatistics();

    res.json({
        success: true,
        data: {
            ingredients: ingredientStats,
            dishes: dishStats,
        },
        timestamp: new Date().toISOString(),
    });
}));

// GET /api/reports/cost-trends - コスト推移
router.get('/cost-trends', asyncHandler(async (req: Request, res: Response) => {
    const db = getDatabase();
    const days = parseInt(req.query.days as string) || 30;

    const costTrends = await db.query(`
        SELECT 
            DATE(created_at) as date,
            AVG(unit_price) as avg_ingredient_cost,
            COUNT(*) as ingredient_count
        FROM ingredients
        WHERE created_at >= DATE_SUB(NOW(), INTERVAL ? DAY)
        GROUP BY DATE(created_at)
        ORDER BY date DESC
        LIMIT 30
    `, [days]);

    const dishTrends = await db.query(`
        SELECT 
            DATE(created_at) as date,
            AVG(total_cost) as avg_dish_cost,
            COUNT(*) as dish_count
        FROM dishes
        WHERE created_at >= DATE_SUB(NOW(), INTERVAL ? DAY)
        GROUP BY DATE(created_at)
        ORDER BY date DESC
        LIMIT 30
    `, [days]);

    const foodTrends = await db.query(`
        SELECT 
            DATE(created_at) as date,
            AVG(total_cost) as avg_food_cost,
            AVG(CASE WHEN price > 0 THEN ((price - total_cost) / price) * 100 ELSE 0 END) as avg_profit_rate,
            COUNT(*) as food_count
        FROM completed_foods
        WHERE created_at >= DATE_SUB(NOW(), INTERVAL ? DAY)
        GROUP BY DATE(created_at)
        ORDER BY date DESC
        LIMIT 30
    `, [days]);

    res.json({
        success: true,
        data: {
            ingredients: costTrends,
            dishes: dishTrends,
            foods: foodTrends,
        },
        timestamp: new Date().toISOString(),
    });
}));

// GET /api/reports/popular-items - 人気アイテム
router.get('/popular-items', asyncHandler(async (req: Request, res: Response) => {
    const popularIngredients = await Ingredient.getPopularIngredients(10);
    const profitableFoods = await CompletedFood.findByProfitRate(10);

    res.json({
        success: true,
        data: {
            popularIngredients,
            profitableFoods,
        },
        timestamp: new Date().toISOString(),
    });
}));

// GET /api/reports/export - レポートエクスポート
router.get('/export', asyncHandler(async (req: Request, res: Response) => {
    const format = req.query.format as string || 'json';
    const type = req.query.type as string || 'summary';

    let data: any = {};

    switch (type) {
        case 'ingredients':
            data = await Ingredient.findAll();
            break;
        case 'dishes':
            data = await Dish.findAll();
            break;
        case 'foods':
            data = await CompletedFood.findAll();
            break;
        default:
            // Summary report
            const ingredientStats = await Ingredient.getGenreStatistics();
            const dishStats = await Dish.getGenreStatistics();
            data = { ingredientStats, dishStats };
    }

    if (format === 'csv') {
        // CSV形式での出力（簡易実装）
        res.setHeader('Content-Type', 'text/csv');
        res.setHeader('Content-Disposition', `attachment; filename="${type}_report.csv"`);
        
        // CSV変換（実際の実装では専用ライブラリを使用）
        const csvData = Array.isArray(data) 
            ? data.map(item => Object.values(item).join(',')).join('\n')
            : JSON.stringify(data);
        
        res.send(csvData);
    } else {
        res.json({
            success: true,
            data,
            exportType: type,
            format,
            timestamp: new Date().toISOString(),
        });
    }

    logger.info(`Exported ${type} report in ${format} format`);
}));

export default router;

File: cooking-cost-system/backend/api/controllers/reports.py
from flask import Blueprint, request, Response
from datetime import datetime, timedelta, timezone
from sqlalchemy import func
from api.database import db
from api.models.item import Item, ItemRelation
from api.utils.response import success, error
from api.utils.auth import require_auth
import csv
import io
import json

reports_bp = Blueprint('reports', __name__)


# GET /api/reports/dashboard
@reports_bp.route('/dashboard', methods=['GET'])
@require_auth
def dashboard():
    total_ingredients = Item.query.filter_by(item_type=1).count()
    total_preps = Item.query.filter_by(item_type=2).count()
    total_foods = Item.query.filter_by(item_type=3).count()

    avg_cost_rate_row = db.session.query(
        func.avg(Item.unit_price)
    ).filter(Item.item_type == 3).scalar()
    avg_cost_rate = round(float(avg_cost_rate_row or 0), 2)

    recent_items = Item.query.order_by(Item.created_at.desc()).limit(5).all()
    type_labels = {1: '食材', 2: '仕込み品', 3: 'お品'}
    recent_activity = [{
        'id': item.id,
        'type': 'ingredient' if item.item_type == 1 else ('prep' if item.item_type == 2 else 'food'),
        'message': f'{type_labels[item.item_type]}「{item.name}」を登録',
        'timestamp': item.created_at.isoformat() + 'Z' if item.created_at else None,
    } for item in recent_items]

    return success({
        'summary': {
            'totalIngredients': total_ingredients,
            'totalPreps': total_preps,
            'totalFoods': total_foods,
            'avgCostRate': avg_cost_rate,
        },
        'recentActivity': recent_activity,
    })


# GET /api/reports/genre-stats
@reports_bp.route('/genre-stats', methods=['GET'])
@require_auth
def genre_stats():
    rows = db.session.query(
        func.coalesce(Item.genre, 'other').label('genre'),
        func.count(Item.id).label('count'),
        func.round(func.avg(Item.unit_price), 4).label('avg_unit_price')
    ).filter(Item.item_type == 1).group_by(
        func.coalesce(Item.genre, 'other')
    ).order_by(func.count(Item.id).desc()).all()

    data = [{'genre': r.genre, 'count': r.count, 'avg_unit_price': float(r.avg_unit_price or 0)} for r in rows]
    return success(data)


# GET /api/reports/cost-trends
@reports_bp.route('/cost-trends', methods=['GET'])
@require_auth
def cost_trends():
    days = max(7, min(request.args.get('days', 30, type=int), 365))
    since = datetime.now(timezone.utc) - timedelta(days=days)

    rows = db.session.query(
        func.date(Item.created_at).label('date'),
        func.round(func.avg(Item.price), 2).label('avg_cost'),
        func.count(Item.id).label('item_count')
    ).filter(
        Item.item_type == 1,
        Item.created_at >= since
    ).group_by(func.date(Item.created_at)).order_by(func.date(Item.created_at)).all()

    data = [{'date': str(r.date), 'avg_cost': float(r.avg_cost or 0), 'item_count': r.item_count} for r in rows]
    return success(data)


# GET /api/reports/popular-items
@reports_bp.route('/popular-items', methods=['GET'])
@require_auth
def popular_items():
    rows = db.session.query(
        Item.id,
        Item.name,
        Item.item_type,
        func.count(ItemRelation.id).label('usage_count')
    ).join(ItemRelation, Item.id == ItemRelation.child_item_id
    ).filter(Item.item_type.in_([1, 2])
    ).group_by(Item.id, Item.name, Item.item_type
    ).order_by(func.count(ItemRelation.id).desc()
    ).limit(10).all()

    data = [{'id': r.id, 'name': r.name, 'item_type': r.item_type, 'usage_count': r.usage_count} for r in rows]
    return success(data)


# GET /api/reports/export
@reports_bp.route('/export', methods=['GET'])
@require_auth
def export_data():
    export_type = request.args.get('type')
    export_format = request.args.get('format', 'json')

    if export_type not in ('ingredients', 'preps', 'foods', 'summary'):
        return error('VALIDATION_ERROR', 'type は ingredients / preps / foods / summary のいずれかを指定してください')

    type_map = {'ingredients': 1, 'preps': 2, 'foods': 3}
    if export_type == 'summary':
        data = {
            'totalIngredients': Item.query.filter_by(item_type=1).count(),
            'totalPreps': Item.query.filter_by(item_type=2).count(),
            'totalFoods': Item.query.filter_by(item_type=3).count(),
        }
        if export_format == 'csv':
            output = io.StringIO()
            w = csv.writer(output)
            w.writerow(data.keys())
            w.writerow(data.values())
            return Response(output.getvalue(), mimetype='text/csv',
                            headers={'Content-Disposition': 'attachment; filename=summary.csv'})
        return success(data)

    items = Item.query.filter_by(item_type=type_map[export_type]).all()
    if export_format == 'csv':
        output = io.StringIO()
        w = csv.writer(output)
        w.writerow(['id', 'name', 'store', 'price', 'quantity', 'unit', 'unit_price', 'genre', 'description'])
        for item in items:
            w.writerow([item.id, item.name, item.store, item.price, item.quantity,
                        item.unit, item.unit_price, item.genre, item.description])
        return Response(output.getvalue(), mimetype='text/csv',
                        headers={'Content-Disposition': f'attachment; filename={export_type}.csv'})

    return success([i.to_dict() for i in items])
